### PR TITLE
fix: support jest config function

### LINKED ISSUE #126: [FR]: Support Jest config returning function
### What is the current behavior?

Looking at the [jest_config_template.mjs](https://github.com/aspect-build/rules_jest/blob/f3d27a1ae5b39e2ea93c23fcfa710d145e4d7783/jest/private/jest_config_template.mjs#L83), seems like it assumes when config is not `json`, it's a module that returns object. If I use config file like described in [Jest doc](https://jestjs.io/docs/configuration), it'll evaluate to `undefined`, which effectively not using my config at all.

```js
/** @returns {Promise<import('jest').Config>} */
module.exports = async () => {
  return {
    verbose: true,
  };
};
```

### Describe the feature

Support `jest.config.js` that exports function / async function returning object. This will be useful for workspace that contains `next.js` project for example.

There's a `next.js` sample in https://github.com/aspect-build/bazel-examples/tree/main/next.js, I haven't run it, but when I tried in my project, it won't use my `jest.config.js`

```js
// Based on https://github.com/vercel/next.js/blob/28454c6ddbc310419467e5415aee26e48d079b46/examples/with-jest/jest.config.js

const nextJest = require('next/jest');

const createJestConfig = nextJest({
  // Provide the path to your Next.js app to load next.config.js and .env files in your test environment.
  // This path is relative to the WORKSPACE file when testing under Bazel.
  dir: './apps/alpha',
});

// Add any custom config to be passed to Jest
const customJestConfig = {
  moduleNameMapper: {
    // Handle module aliases (this will be automatically configured for you soon)
    '^@/components/(.*)$': '<rootDir>/components/$1',

    '^@/pages/(.*)$': '<rootDir>/pages/$1',
  },
  testEnvironment: 'jest-environment-jsdom',
};

// createJestConfig is exported this way to ensure that next/jest can load the Next.js config which is async
module.exports = createJestConfig(customJestConfig);
```

### PR DESCRIPTION
**Pull request by panjiesw**
fixes #126 

**Type of change**

- [x] Bug fix (change which fixes an issue)
- [ ] New feature or functionality (change which adds functionality)
- [ ] Style (white-space, formatting, etc...)
- [ ] Refactor (a code change that neither fixes a bug or adds a new feature)
- [ ] Performance (a code change that improves performance)
- [ ] Documentation (updates to documentation or READMEs)
- [ ] Chore (any other change that doesn't affect source or test files)

**For changes visible to end-users**

- [ ] Breaking change (this change will force users to change their own code or config)
- [ ] Relevant documentation has been updated
- [ ] Suggested release notes are provided below:

**Test plan**

How has this been tested?

- [ ] Covered by existing test cases
- [x] New test cases added
- [ ] Manual testing

If part of the test plan included manual testing, please provide instructions so we can reproduce.
Please also list any relevant details for your test configuration.

### CODE REVIEW

File: jest/tests/case11/case11.index.js [hunk 1]
```diff
@@ -0,0 +1,5 @@
+function foobar() {
```
Comment (jbedard) [inline jest/tests/case11/case11.index.js:1]: Do we need this file to test it? I think we just need the `.test.js` and 2 config files?

Comment (panjiesw) [reply jest/tests/case11/case11.index.js:1]: @jbedard I followed other test cases, but you're right, we only need the `.test.js` files

Comment (panjiesw) [reply jest/tests/case11/case11.index.js]: @jbedard applied in [`57d1180` (#127)](https://github.com/aspect-build/rules_jest/pull/127/commits/57d1180be66517ecc7c0be3e511134bfeeddc5e3)

Comment (jbedard) [reply jest/tests/case11/case11.index.js]: 👍

**Comment by jbedard** [comment]
Awesome 👍 🎉

Anything still in progress or is it good to go?

**Comment by panjiesw** [comment]
@jbedard no pending changes, good to go from me 🙂

**Review by jbedard — APPROVED** [review]

**Comment by panjiesw** [comment]
@jbedard any chance of a new release that will include this soon?

**Comment by jbedard** [comment]
Done 👍

**Pull request merged**